src/Microsoft.DotNet.Build.CloudTestTasks/DownloadFromAzure.cs

```diff
@@ -95,10 +96,16 @@ public override bool Execute()
                         {
                             blobDirectory = blob.Substring(0, dirIndex);
                             blobFilename = blob.Substring(dirIndex + 1);
-                        }
-                        if(BlobNamePrefix != null)
-                        {
-                            blobDirectory = blobDirectory.Substring(BlobNamePrefix.Length);
+
+                            // Trim blob name prefix (directory part) from download to blob directory
+                            if(BlobNamePrefix != null)
+                            {
+                                if(BlobNamePrefix.Length > dirIndex)
+                                {
+                                    BlobNamePrefix = BlobNamePrefix.Substring(0, dirIndex);
```

Comment:
The prefix is used for two different things here.

1. Only return me the set of blobs which start with this prefix
2. Trim the irrelevant directory structure.

    This task (DownloadFromAzure) is used in a variety of ways.  Many repos (coreclr, core-setup), have blobs in a structure like `[container name]/[folder structure]/[blobs I care about]`.  So, the folder structure is used to determine the piece that I care about, but they don't want to bring down that entire structure locally.  I'm not prescribing best practice, I'm just outlining how repo's are currently utilizing the tools.  ie, They want to download `coreclr-beta-version/Release/pkgs`, locally into a folder named `packages/AzureTransfer`, not locally into a folder named `packages/AzureTransfer/Release/pkgs`.  Thus, the `BlobNamePrefix` is dual purposed in Azure to say, give me all of the files from a container under a certain folder and starting with a certain filename, and put them into such and such folder (this is the case with BlobNamePrefix="Release/pkgs/Microsoft").  In other scenarios,  the BlobNamePrefix might be "Release/pkgs", and that means give me everything in a container under "Release/pkgs".

Anyways, there's a variety of ways this could be handled.  The fix I'm proposing, I believe is the most compatible with all repo's.

I agree with, (2), but I'd assume that coreclr opted for this method because they didn't want to pull down the transport packages.  Whether they should or shouldn't is a different question and would depend on if we decide to prescribe stricter regulations about how to use this task.